src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -387,33 +387,12 @@ public void SetStateMachine(IAsyncStateMachine stateMachine)
         {
             IAsyncStateMachineBox box = GetStateMachineBox(ref stateMachine);
 
-            // TODO https://github.com/dotnet/coreclr/issues/12877:
-            // Once the JIT is able to recognize "awaiter is ITaskAwaiter" and "awaiter is IConfiguredTaskAwaiter",
-            // use those in order to a) consolidate a lot of this code, and b) handle all Task/Task<T> and not just
-            // the few types special-cased here.  For now, handle common {Configured}TaskAwaiter.  Having the types
-            // explicitly listed here allows the JIT to generate the best code for them; otherwise we'll fall through
-            // to the later workaround.
-            if (typeof(TAwaiter) == typeof(TaskAwaiter) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<object>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<string>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<byte[]>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<bool>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<byte>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<int>) ||
-                typeof(TAwaiter) == typeof(TaskAwaiter<long>))
+            if (awaiter is ITaskAwaiter)
```

Comment:
WOO HOO!

Reply:
I ended up wrapping these new `is` tests in a value class guard, otherwise cases where the arguments were byref-to-ref classes still ran through these interface tests -- take a look and make sure the new logic is reasonable.